### ISSUE TITLE
smartos.vm_present failing to accept customer_metadata

### ISSUE THREAD

**Issue by garethhowell** [saltstack/salt#51570]
### Description of Issue/Question
I’m trying to use salt to orchestrate the creation and configuration of zones on my home SmartOS servers.

I’m using @sjorge smartos.vm_present module to create the KVM using the latest Ubuntu 18.04 zvol image. I then want to use salt-cloud and saltify to auto-install the salt-minion and connect it to the salt-master.

Saltify requires root access into the VM, but Ubuntu disables ssh logins for the root account and the default `ubuntu` account has no sudo privileges, so I tried including salt-cloud meta data to the state in the “customer_metadata” key to to create a user with sudo rights.

The state fails with 

`Comment: {u'bad_values': [u'set_customer_metadata'], u'bad_properties': [], u'missing_properties': []}`

### Setup
```
ac99517a-72ac-44c0-90e6-c7ce3d944a0a:
  smartos.image_present

docker1.agdon.net:
  smartos.vm_present:
    - vmconfig:
        brand: kvm
        alias: docker1
        vcpus: 1
        autoboot: true
        ram: 4096
        resolvers:
          - 172.29.12.7
          - 8.8.8.8
        tags:
          label: 'docker1'
          owner: 'garethhowell'
        disks:
          disk0:
            image_uuid: ac99517a-72ac-44c0-90e6-c7ce3d944a0a
            model: virtio
            boot: true
          disk1:
            size: 32000
            model: virtio
            compression: lz4
        nics:
          "82:1b:33:49:e9:15":
            nic_tag: admin
            model: virtio
            mtu: 1500
            ip: 172.29.12.34
            netmask: 255.255.255.0
            gateway: 172.29.12.1
            allow_ip_spoofing: true
          "82:1b:33:49:e9:16":
            nic_tag: dmz
            mtu: 1500
            model: virtio
            ip: 172.29.13.34
            netmask: 255.255.255.0
            gateway: 172.29.13.1
            allow_ip_spoofing: true
        customer_metadata:
          root_authorized_keys:
            - "ssh-rsa XXXX root@salt"
          users:
            - default
            - name: gareth
              lock_passwd: False
              ssh_import_id: gareth
              sudo: "ALL=(ALL) NOPASSWD:ALL"
              shell: /bin/bash
              ssh_authorized_keys:
                - "ssh-rsa XXX"
```
### Steps to Reproduce Issue
`sudo salt global_deneb state.apply kvm_docker1`

### Versions Report
```
Salt Version:
           Salt: 2019.2.0-566-g79217f3
 
Dependency Versions:
           cffi: 1.11.5
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.19
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.15 (default, Jan 28 2019, 05:17:09)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.1.2
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 5.1.1
            ZMQ: 4.2.5
 
System Versions:
           dist:   
         locale: UTF-8
        machine: i86pc
        release: 5.11
         system: SunOS
        version: Not Installed
 ```

**Comment by sjorge** [comment]
If you use -l debug it should print the json file it generated, on failure it does not clean them up. It would be interesting to look at what it generates.

Also based on vmadm manpage it looks like it wants a simple key-value pair for customer_metadata:

```
       customer_metadata:

           This field allows metadata to be set and associated with this VM. The
           value should be an object with only top-level key=value pairs.

           NOTE1: for historical reasons, do not put keys in here that match the
           pattern *_pw. Those keys should go in internal_metadata instead.

           NOTE2: keys that are prefixed with one of the prefixes listed in
           internal_metadata_namespaces will not be read from customer_metadata but
           rather from internal_metadata. These will also be read-only from within
           the zone.

           type: JSON Object (key: value)
           vmtype: ANY
           listable: no
           create: yes
           update: yes (but see special notes on update command)
           default: {}
```

The following works fine for me:

```
        customer_metadata:
          root_authorized_keys: "ssh-type hash comment"
```

**Comment by sjorge** [comment]
Also the wiki seems to support that:
https://wiki.smartos.org/display/DOC/How+to+create+a+KVM+VM+%28+Hypervisor+virtualized+machine+%29+in+SmartOS

Also look at the section mentioning cloud-init, the data needs to be properly encoded given this seems to be a simple key-value for vmadm.

**Comment by sjorge** [comment]
Using the proper encoding SmartOS's vmadm expect works fine.

```
[root@carbon /zdata/cronos/salt/states]# vim test.sls
[root@carbon /zdata/cronos/salt/states]# salt-call state.sls test
local:
----------
          ID: openhab
    Function: smartos.vm_present
      Result: True
     Comment: vm openhab updated
     Started: 17:56:28.227729
    Duration: 4585.077 ms
     Changes:
              ----------
              openhab:
                  ----------
                  set_customer_metadata:
                      ----------
                      cloud-init:user-data:
                          users:
                            - default
                            - name: sjorge
                               lock_passwd: False
                               sudo: sudo: "ALL=(ALL) NOPASSWD:ALL"
                               shell: /bin/bash
                               ssh_authorized_keys:
                                 - "<key removed>"
```

```yaml
        customer_metadata:
          root_authorized_keys: "<key removed>"
          "cloud-init:user-data": "users:\n  - default\n  - name: sjorge\n     lock_passwd: False\n     sudo: sudo: \"ALL=(ALL) NOPASSWD:ALL\"\n     shell: /bin/bash\n     ssh_authorized_keys:\n       - \"<key removed>"\""
```

**Comment by sjorge** [comment]
This also seems to work:

```
        customer_metadata:
          root_authorized_keys: "<key>"
          cloud-init:user-data: |
            users:
              - default
              - name: sjorge
                lock_passwd: False
                sudo: "ALL=(ALL) NOPASSWD:ALL"
                shell: /bin/bash
                ssh_authorized_keys:
                  - "<key>"
```

I have not tested if cloud-init is happy with it, but SmartOS is atleast.

**Comment by garethhowell** [comment]
Thanks for all that, Jorge.
I think I over-interpreted the man pages. As you say, the wiki and the
vmadm man page refer to json encoded strings, but the salt-cloud pages
refer to yaml. I interpreted that as meaning that as I was doing this from
salt, I could put the yaml in the state.

I’ll encode the yaml and see what happens.

Gareth

**Comment by garethhowell** [comment]
OK, I modified the state to have the customer_metadata encoded and the state now passes. However, I can't login as `gareth` or `ubuntu`.

I repeated the process using a json file to `vmadm create -f` and the user `gareth` does not get created. Luckily, the user `ubuntu` does get created, so I can login and interrogate `/var/logcloud-init-log` for clues.

Root's `authorized_keys` does get amended and the default user gets created so this indicates a problem with the `cloud-init` string somewhere. I need to resolve this using json files before I revert back to the salt state file.

**Comment by garethhowell** [comment]
If I create a kvm using the following:
```
{
  "alias": "docker1",
  "hostname": "docker1.agdon.net",
  "brand": "kvm",
  "vcpus": 1,
  "autoboot": true,
  "ram": 4096,
  "resolvers": [ 
     "172.29.12.7",
      "8.8.8.8"
  ],
  "disks": [
    {
      "boot": true,
      "image_uuid": "ac99517a-72ac-44c0-90e6-c7ce3d944a0a",
      "model": "virtio"
    }
  ],
  "nics": [
    {
      "nic_tag": "admin",
      "model": "virtio",
      "ip": "172.29.12.34",
      "netmask": "255.255.255.0",
      "gateway": "172.29.12.1",
      "primary": true,
      "allow_ip_spoofing": true
    },
    {
      "nic_tag": "dmz",
      "model": "virtio",
      "ip": "172.29.13.34",
      "netmask": "255.255.255.0",
      "gateway": "172.29.13.1",
      "allow_ip_spoofing": true
    }
  ],
  "customer_metadata" : {
    "root_authorized_keys" : "ssh-rsa XXX",
    "cloud-init:user-data": "users:\n  - default\n  - name: gareth\n    ssh_import_id: garethhowell\n    lock_passwd: false\n    sudo: \"ALL=(ALL) NOPASSWD:ALL\"\n    shell: /bin/bash"
  }
}
```
I can login as the default user `ubuntu`. However, there are a couple of odd things:
* the user `gareth` does not get created
* the user `ubuntu` has `garethhowell`'s ssh key in its ``authorized_keys`

It's almost as if cloud-init merged the two users together in some way.

If I move the customer_metadata into the salt state as follows:
```
docker1.agdon.net:
  smartos.vm_present:
    - vmconfig:
        brand: kvm
        alias: docker1
        vcpus: 1
        autoboot: true
        ram: 4096
        resolvers:
          - 172.29.12.7
          - 8.8.8.8
        tags:
          label: 'docker1'
          owner: 'garethhowell'
        disks:
          disk0:
            image_uuid: ac99517a-72ac-44c0-90e6-c7ce3d944a0a
            model: virtio
            boot: true
          disk1:
            size: 32000
            model: virtio
            compression: lz4
        nics:
          "82:1b:33:49:e9:15":
            nic_tag: admin
            model: virtio
            mtu: 1500
            ip: 172.29.12.34
            netmask: 255.255.255.0
            gateway: 172.29.12.1
            allow_ip_spoofing: true
          "82:1b:33:49:e9:16":
            nic_tag: dmz
            mtu: 1500
            model: virtio
            ip: 172.29.13.34
            netmask: 255.255.255.0
            gateway: 172.29.13.1
            allow_ip_spoofing: true
        customer_metadata:
          root_authorized_keys: "ssh-rsa XXX"
          cloud-init:user-data: "users:\n  - default\n  - name: gareth\n    ssh_import_id: garethhowell\n    lock_passwd: false\n    sudo: \"ALL=(ALL) NOPASSWD:ALL\"\n    shell: /bin/bash"
```
The VM gets created but I cannot login as `ubuntu`. So there is definitely a difference in behaviour between the two. I've put the debug output from the salt run at https://pastebin.com/SWYrat7b

**Comment by sjorge** [comment]
To get the useful debug output you need to run the state via salt-call on the minion. ```salt-call -l info state.sls your_state_here``` (or `-l debug` if info was not helpful).
Does `vmadm get UUID` report a any difference for the could-init:user-data field between a vm create using salt or manually?

**Comment by zsoltf** [comment]
I had the same issue, where the `root_authorized_keys` was picked up but anything under `cloud-init:user-data` got ignored.

Turns out the `root_authorized_keys` is meant as an easy alternative to cloud-init for injecting ssh keys. It adds the keys to the root user and the first user. When combined with cloud-init, the root and ubuntu user get the root_authorized_keys.

`cloud-init:user-data` does get populated in the vm, it should be visible by running `mdata-get cloud-init:user-data` as root. According to the cloud-init logs, the user-data was loaded but there were no errors. After some digging, I found that cloud-init expects user-data to start with `#cloud-config`

https://wiki.smartos.org/how-to-create-an-hvm-zone/#passing-ssh-keys-to-the-vm
https://cloudinit.readthedocs.io/en/latest/topics/datasources/smartos.html 

TLDR;

`cloud-init:user-data` should start with `#cloud-config`. Only add `ssh_authorized_keys` if you plan on using the default user.

        customer_metadata:
          root_authorized_keys: "ssh-rsa root_and_ubuntus_key"
          cloud-init:user-data: |
            #cloud-config

            users:
              - default
              - name: zsolt
                lock_passwd: False
                shell: /bin/bash
                ssh_authorized_keys:
                  - ssh-rsa mykey

Hope this helps.

**Comment by garethhowell** [comment]
OK, very interesting. I'll dig deeper.

Apologies to @sjorge for dropping the ball after his initial response.

**Comment by garethhowell** [comment]
I seem to have solved this. I used the following in a salt state
```
        customer_metadata:
          root_authorized_keys: "ssh-rsa blah-blah-blah"
          cloud-init:user-data: |
            #cloud-config
            hostname: ubuntu_test1
            domain: agdon.net
            searchdomains:
              - agdon.net

            users:
              - default
              - name: guff
                passwd: "blah-blah-blah"
                lock_passwd: False
                shell: /bin/bash
                ssh_import_id: guff
                sudo: ALL=(ALL) NOPASSWD:ALL
            runcmd:
              - echo ubuntu_test1 > /etc/hostname
              - mkdir /run/salt
              - curl -L "https://bootstrap.saltstack.com" -o /run/salt/bootstrap-salt.sh
              - [ sh, /run/salt/bootstrap-salt.sh, -U, -i, ubuntu_test1, -A, 172.29.12.22, git ]

```
This executes correctly and the end result is a running Ubuntu KVM with a running salt minion.
